lib/server.js

```diff
@@ -145,51 +150,68 @@ module.exports = function createServer(config, log) {
             }
 
             var retryAfter = [blockEmail, blockIpEmail, blockIp].reduce(max)
+            var block = retryAfter > 0
+            var suspect = false
 
-            var allowedIP = ip in allowedIPs.ips
-            var allowedEmailDomain = allowedEmailDomains.isAllowed(email)
-
-            if (retryAfter > 0) {
-              if (allowedIP) {
-                retryAfter = 0
-              }
-              if (allowedEmailDomain) {
-                retryAfter = 0
-              }
+            if (requestChecks.treatEveryoneWithSuspicion) {
+              suspect = true
             }
+            // The private branch puts some additional request checks here.
+            // We just use the variables so that eslint doesn't complain about them.
+            payload || headers
+
             var canUnblock = emailRecord.canUnblock()
 
             // IP's that are in blocklist should be blocked
             // and not return a retryAfter because it is not known
             // when they would be removed from blocklist
-            if (config.ipBlocklist.enable && blockListManager.contains(ip) && !allowedIP && !allowedEmailDomain) {
+            if (config.ipBlocklist.enable && blockListManager.contains(ip)) {
               if (!config.ipBlocklist.logOnly) {
-
-                // When blocklist is enabled, explicitly skip setting records.
-                // Requests issued while on blocklist will not count towards
-                // the rate limit.
-                return {
-                  block: true,
-                  unblock: canUnblock
-                }
+                block = true
+                retryAfter = 0
               }
             }
 
             return setRecords(email, ip, emailRecord, ipRecord, ipEmailRecord)
               .then(
                 function () {
                   return {
-                    block: retryAfter > 0,
+                    block: block,
+                    retryAfter: retryAfter,
                     unblock: canUnblock,
-                    retryAfter: retryAfter
+                    suspect: suspect
                   }
                 }
               )
           }
         )
         .then(
           function (result) {
-            log.info({ op: 'request.check', email: email, ip: ip, action: action, block: result.block })
+            // Regardless of all the above checks, there are some
+            // IPs and emails that we just won't block.  These are
+            // typically for Mozilla QA purposes.  We check them at
+            // the end so as not to pay the overhead of the checks
+            // on the many requests that are *not* QA-related.
```

Comment:
The checking of `allowedIPs` and `allowedEmailDomains` has been pulled out into a separate function, so that we can be sure it's always executed regardless of what happens in the logic above.